File: common/WebGL/Misc/SceneBase.js
// Définition de la classe SceneBase


// classes et modules nécessaires
Requires("libs/Process/FrameBufferObject.js");


/**
 * Définit une base pour dessiner une scène
 */
class SceneBase
{
    /**
     * Définit une base pour dessiner une scène
     * @param deferredShading : true si cette classe gère le FBO m_GBuffer pour un dessin en deux passes
     */
    constructor(deferredShading)
    {
        // configurer les modes de dessin
        gl.enable(gl.DEPTH_TEST);
        gl.enable(gl.CULL_FACE);

        // couleur du fond
        gl.clearColor(0.4, 0.4, 0.4, 0.0);

        // lampes
        this.m_Lights = [];

        // FBO pour dessiner en mode Multiple Render Targets
        this.m_DeferredShading = deferredShading;
        this.m_GBuffer = null;     // sera initialisé dans onSurfaceChanged

        // matrice de projection et de transformation
        this.m_Mat4Projection = mat4.create();
        this.m_Mat4ModelView = mat4.create();
    }


    /** destructeur */
    destroy()
    {
        if (this.m_GBuffer != null) this.m_GBuffer.destroy();
    }


    /** affiche les informations sur la caméra **/
    updateInfos()
    {
        throw "method SceneBase.updateInfos should be overloaded in Scene subclass";
    }


    /**
     * appelée quand la taille de la vue OpenGL change
     * @param width : largeur en nombre de pixels de la fenêtre
     * @param height : hauteur en nombre de pixels de la fenêtre
     * @param scale : facteur d'agrandissement du FBO pour simuler un antialiasing global
     */
    onSurfaceChanged(width, height, scale=2)
    {
        // met en place le viewport
        gl.viewport(0, 0, width, height);

        // créer un FBO pour dessiner hors écran avec plusieurs buffers pour stocker toutes les informations nécessaires
        if (this.m_GBuffer != null) this.m_GBuffer.destroy();
        if (this.m_DeferredShading) {
            this.m_GBuffer = new FrameBufferObject(width*scale, height*scale, gl.TEXTURE_2D, gl.TEXTURE_2D, 3);
        } else {
            this.m_GBuffer = null;
        }
    }


    /**
     * appelée quand on enfonce le bouton du milieu et qu'on déplace la souris
     * @param dd : nombre de pixels parcourus
     */
    onZoom(dd)
    {
    }


    /**
     * appelée quand on enfonce le bouton droit et qu'on déplace la souris
     * @param dx : nombre de pixels parcourus horizontalement
     * @param dy : nombre de pixels parcourus verticalement
     */
    onPan(dx, dy)
    {
    }


    /**
     * appelée quand on enfonce le bouton gauche et qu'on déplace la souris
     * @param oldX : ancienne absisse de la souris
     * @param oldY : ancienne ordonnée de la souris
     * @param x : nouvelle absisse de la souris
     * @param y : nouvelle ordonnée de la souris
     * @param width : largeur de la vue OpenGL
     * @param height : hauteur de la vue OpenGL
     */
    onRotate(oldX, oldY, x, y, width, height)
    {
    }


    /**
     * appelée quand on appuie sur une touche du clavier
     * @param code : touche enfoncée
     */
    onKeyDown(code)
    {
        switch (code) {
        case 'P'.charCodeAt(0):
            // mettre en pause
            PauseTime();
            break;
        }
    }


    /**
     * retourne la matrice ModelView pour dessiner la scène
     * @return matrice ModelView
     */
    getModelView()
    {
        return this.m_Mat4ModelView;
    }


    /**
     * ajoute une lampe pour éclairer le matériau du renderer
     * @param light : lampe à ajouter
     */
    addLight(light)
    {
        this.m_Lights.push(light);
    }


    /**
     * transforme toutes les lampes
     * c'est à dire calcule la position relative à la caméra m_PositionCamera
     * à partir de la position relative à la scène m_Position
     * @param mat4View matrice de positionnement de la scène par rapport à la caméra
     */
    transformLights(mat4View)
    {
        for (let light of this.m_Lights) {
            light.transform(mat4View);
        }
    }


    /**
     * dessine les shadow maps des lampes
     * @param mat4View : matrice de positionnement de la scène par rapport à la caméra
     */
    makeShadowMaps(mat4View)
    {
        for (let light of this.m_Lights) {
            // calculer sa shadow map (si la lampe en gère une)
            light.makeShadowMap(this, mat4View);
        }
    }


    /**
     * transforme les lampes puis calcule leur shadow map s'il y en a
     * @param mat4View : matrice de positionnement de la scène par rapport à la caméra
     */
    prepareLights(mat4View)
    {
        this.transformLights(mat4View);
        this.makeShadowMaps(mat4View);
    }


    /**
     * appelée pour dessiner en mode MRT
     * @param mat4Projection : matrice de projection
     * @param mat4ModelView : matrice de vue
     */
    drawDeferredShading(mat4Projection, mat4ModelView)
    {
        // rediriger les dessins vers le FBO
        if (this.m_DeferredShading && this.m_GBuffer != null) this.m_GBuffer.enable();

        // effacer l'écran
        gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);

        // dessiner les objets de la scène
        this.onDraw(mat4Projection, mat4ModelView);

        // revenir au dessin sur l'écran
        if (this.m_DeferredShading && this.m_GBuffer != null) this.m_GBuffer.disable();

        // (debug) dessiner l'un des buffers sur l'écran
        //this.m_GBuffer.onDraw(gl.COLOR_ATTACHMENT0);
        //this.m_GBuffer.onDraw(gl.DEPTH_ATTACHMENT);
    }


    /**
     * rajoute l'éclairement d'une lampe
     * @param light : lampe à ajouter
     * @param is_first : mettre true si c'est la première qu'on ajoute ainsi, false sinon
     */
    addLighting(light, is_first=false)
    {
        if (!this.m_DeferredShading || this.m_GBuffer == null) return;

        if (is_first) {

            // dessiner l'éclairement de la première lampe
            light.process(this.m_GBuffer);

        } else {

            // passer en mode blending, en tenant compte du canal alpha des éclairements ajoutés
            gl.enable(gl.BLEND);
            gl.blendFunc(gl.SRC_ALPHA, gl.SRC_ALPHA);

            // rajouter l'éclairement de la lampe
            light.process(this.m_GBuffer);

            // revenir en mode normal
            gl.disable(gl.BLEND);
        }
    }


    /**
     * rajoute les éclairements de toutes les lampes
     */
    addLightings()
    {
        if (this.m_DeferredShading && this.m_GBuffer != null) {
            let is_first = true;
            for (let light of this.m_Lights) {
                this.addLighting(light, is_first);
                is_first = false;
            }
        }
    }


    /**
     * dessine les objets de la scène
     * @param mat4Projection : matrice de projection
     * @param mat4ModelView : matrice de vue
     */
    onDraw(mat4Projection, mat4ModelView)
    {
        throw "method SceneBase.onDraw should be overloaded in Scene subclass";
    }


    /**
     * dessin de la scène sur l'écran
     */
    onDrawFrame()
    {
        // effacer l'écran
        gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);

        // construire la matrice MV
        let mat4ModelView = this.getModelView();

        // positionner les lampes et dessiner leur shadow map s'il y en a
        this.prepareLights(mat4ModelView);

        // dessiner les objets dans le FBO mrt (ou sur l'écran)
        this.drawDeferredShading(this.m_Mat4Projection, mat4ModelView);

        // ajouter les éclairements des lampes
        this.addLightings();
    }
}

File: common/C++/Misc/SceneBase.cpp
// Définition de la classe SceneBase

#include <GL/glew.h>
#include <GL/gl.h>

#include <iostream>
#include <ios>

#include <utils.h>
#include <SceneBase.h>



/**
 * Définit une base pour dessiner une scène
 * @param deferredShading : true si cette classe gère le FBO m_GBuffer pour un dessin en deux passes
 */
SceneBase::SceneBase(bool deferredShading)
{
    // configurer les modes de dessin
    glEnable(GL_DEPTH_TEST);
    glEnable(GL_CULL_FACE);

    // couleur du fond
    glClearColor(0.4, 0.4, 0.4, 0.0);

    // FBO pour dessiner en mode Multiple Render Targets
    m_DeferredShading = deferredShading;
    m_GBuffer = nullptr;     // sera initialisé dans onSurfaceChanged

    // matrice de projection et de transformation
    m_Mat4Projection = mat4::create();
    m_Mat4ModelView = mat4::create();
}


/** destructeur */
SceneBase::~SceneBase()
{
    if (m_GBuffer != nullptr) delete m_GBuffer;
}


/** affiche les informations sur la caméra **/
void SceneBase::updateInfos()
{
}


/**
 * appelée quand la taille de la vue OpenGL change
 * @param width : largeur en nombre de pixels de la fenêtre
 * @param height : hauteur en nombre de pixels de la fenêtre
 * @param scale : facteur d'agrandissement du FBO pour simuler un antialiasing global
 */
void SceneBase::onSurfaceChanged(int width, int height, int scale)
{
    // met en place le viewport
    glViewport(0, 0, width, height);

    // créer un FBO pour dessiner hors écran avec plusieurs buffers pour stocker toutes les informations nécessaires
    if (m_GBuffer != nullptr) delete m_GBuffer;
    if (m_DeferredShading) {
        m_GBuffer = new FrameBufferObject(width*scale, height*scale, GL_TEXTURE_2D, GL_TEXTURE_2D, 3);
    } else {
        m_GBuffer = nullptr;
    }
}


/**
 * appelée quand on enfonce le bouton du milieu et qu'on déplace la souris
 * @param dz : nombre de pixels parcourus
 */
void SceneBase::onZoom(float dz)
{
}


/**
 * appelée quand on enfonce le bouton droit et qu'on déplace la souris
 * @param dx : nombre de pixels parcourus horizontalement
 * @param dy : nombre de pixels parcourus verticalement
 */
void SceneBase::onPan(float dx, float dy)
{
}


/**
 * appelée quand on enfonce le bouton gauche et qu'on déplace la souris
 * @param oldX : ancienne absisse de la souris
 * @param oldY : ancienne ordonnée de la souris
 * @param x : nouvelle absisse de la souris
 * @param y : nouvelle ordonnée de la souris
 * @param width : largeur de la vue OpenGL
 * @param height: hauteur de la vue OpenGL
 */
void SceneBase::onRotate(float oldX, float oldY, float x, float y, int width, int height)
{
}


/**
 * appelée quand on appuie sur une touche du clavier
 * @param code : touche enfoncée
 */
void SceneBase::onKeyDown(unsigned char code)
{
    switch (code) {
    case 'P':
        // mettre en pause
        Utils::PauseTime();
        break;
    default:
        ;
    }
}


/**
 * retourne la matrice ModelView pour dessiner la scène
 * @return matrice ModelView
 */
mat4 SceneBase::getModelView()
{
    return m_Mat4ModelView;
}


/**
 * ajoute une lampe pour éclairer le matériau du renderer
 * @param light : lampe à ajouter
 */
void SceneBase::addLight(Light* light)
{
    m_Lights.push_back(light);
}


/**
 * transforme toutes les lampes
 * c'est à dire calcule la position relative à la caméra m_PositionCamera
 * à partir de la position relative à la scène m_Position
 * @param mat4View matrice de positionnement de la scène par rapport à la caméra
 */
void SceneBase::transformLights(mat4& mat4View)
{
    for (Light* light: m_Lights) {
        light->transform(mat4View);
    }
}


/**
 * dessine les shadow maps des lampes
 * @param mat4View : matrice de positionnement de la scène par rapport à la caméra
 */
void SceneBase::makeShadowMaps(mat4& mat4View)
{
    for (Light* light: m_Lights) {
        // calculer sa shadow map (si la lampe en gère une)
        light->makeShadowMap(this, mat4View);
    }
}


/**
 * transforme les lampes puis calcule leur shadow map s'il y en a
 * @param mat4View : matrice de positionnement de la scène par rapport à la caméra
 */
void SceneBase::prepareLights(mat4& mat4View)
{
    transformLights(mat4View);
    makeShadowMaps(mat4View);
}


/**
 * appelée pour dessiner en mode MRT
 * @param mat4Projection : matrice de projection
 * @param mat4ModelView : matrice de vue
 */
void SceneBase::drawDeferredShading(mat4& mat4Projection, mat4& mat4ModelView)
{
    // rediriger les dessins vers le FBO
    if (m_GBuffer != nullptr) m_GBuffer->enable();

    // effacer l'écran
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

    // dessiner les objets de la scène
    onDraw(mat4Projection, mat4ModelView);

    // revenir au dessin sur l'écran
    if (m_GBuffer != nullptr) m_GBuffer->disable();

    // dessiner l'un des buffers sur l'écran
    //m_GBuffer->onDraw(GL_COLOR_ATTACHMENT0);
    //m_GBuffer->onDraw(GL_DEPTH_ATTACHMENT);
}


/**
 * rajoute l'éclairement d'une lampe
 * @param light : lampe à ajouter
 * @param is_first : mettre true si c'est la première qu'on ajoute ainsi, false sinon
 */
void SceneBase::addLighting(Light* light, bool is_first)
{
    if (!m_DeferredShading || m_GBuffer == nullptr) return;

    if (is_first) {

        // dessiner l'éclairement de la première lampe
        light->process(m_GBuffer);

    } else {

        // passer en mode blending, en tenant compte du canal alpha des éclairements ajoutés
        glEnable(GL_BLEND);
        glBlendFunc(GL_SRC_ALPHA, GL_SRC_ALPHA);

        // rajouter l'éclairement de la lampe
        light->process(m_GBuffer);

        // revenir en mode normal
        glDisable(GL_BLEND);
    }
}


/**
 * rajoute les éclairements de toutes les lampes
 */
void SceneBase::addLightings()
{
    if (m_DeferredShading && m_GBuffer != nullptr) {
        bool is_first = true;
        for (Light* light: m_Lights) {
            addLighting(light, is_first);
            is_first = false;
        }
    }
}


/**
 * dessine les objets de la scène
 * @param mat4Projection : matrice de projection
 * @param mat4ModelView : matrice de vue
 */
void SceneBase::onDraw(mat4& mat4Projection, mat4& mat4ModelView)
{
}


/**
 * dessin de la scène sur l'écran
 */
void SceneBase::onDrawFrame()
{
    // effacer l'écran
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

    // construire la matrice MV
    mat4 mat4ModelView = getModelView();

    // positionner les lampes et dessiner leur shadow map s'il y en a
    prepareLights(mat4ModelView);

    // dessiner les objets dans le FBO mrt (ou sur l'écran)
    drawDeferredShading(m_Mat4Projection, mat4ModelView);

    // ajouter les éclairements des lampes
    addLightings();
}
